Add IBI quality method for ECG and PPG

peterhcharlton opened the pull request:
# Description

This PR aims to add a new quality assessment method for ECG and PPG signals. Briefly, this method assesses signal quality by comparing beat detections from two beat detectors. It outputs a quality array containing 0s (low quality) and 1s (high), where each detected interbeat interval (IBI) is deemed to be either low or high quality.

# Proposed Changes

- I added a `signal_ibiquality()` function, which is suitable for use with ECG and PPG signals
- I updated `ppg_quality()` and `ecg_quality` with this new quality assessment method. When the method is selected, these functions now call `signal_ibiquality`.

# Checklist

Here are some things to check before creating the PR. If you encounter any issues, do let us know :)

- [X] I have read the [CONTRIBUTING](https://github.com/neuropsychology/NeuroKit/blob/master/.github/CONTRIBUTING.rst#structure-and-code) file.
- [X] My PR is targeted at the **dev branch** (and not towards the master branch).
- [ ] I ran the [CODE CHECKS](https://github.com/neuropsychology/NeuroKit/blob/master/.github/CONTRIBUTING.rst#run-code-checks) on the files I added or modified and fixed the errors.
- [ ] I have added the newly added features to **News.rst** (if applicable)

# Notes
This builds on the changes made in [PR1129](https://github.com/neuropsychology/NeuroKit/pull/1129) and [PR1130](https://github.com/neuropsychology/NeuroKit/pull/1130).

peterhcharlton commented on neurokit2/signal/signal_templatequality.py:
```diff
@@ -103,42 +97,37 @@ def _calc_template_morph(signal, beat_inds, signal_type, sampling_rate=1000):
+    for i in range(ind_morph.shape[0]):
+        temp = np.corrcoef(ind_morph.iloc[i], templ_morph)
+        cc[i] = temp[0, 1]
+    beat_inds_for_cc = beat_inds[ind_morph.index.to_numpy() - 1]
```
Perhaps we need to look back at this and work out whether this needs changing. I can't remember whether I made this change (I presume it's a commit in a previous PR, rather than a new commit in this PR).

peterhcharlton commented on neurokit2/signal/signal_templatequality.py:
```diff
@@ -172,30 +160,25 @@
+    dis = np.zeros(ind_morph.shape[0])
+    for i in range(ind_morph.shape[0]):
+        dis[i] = _calc_dis(ind_morph.iloc[i], templ_morph)
+    beat_inds_for_dis = beat_inds[ind_morph.index.to_numpy() - 1]
```
This is a similar issue to that above - again, perhaps we should look back and work out what's best here.

DerAndereJohannes commented:
Hi Pete, thank you for waiting. I took the time to read the papers and go through your implementations of the new features and have separated my comments around each section. Here goes:

# RSP
- `_rsp_findpeaks_schafer`: Based on the implementation that is written, it seems to follow the Bettermann et al. `Count-orig` algorithm more than the `Count-adv` method that is proposed by Schaefer. I mostly deduct this by your use of the 0.2 constants. Do correct me if I am wrong.
- `_rsp_findpeaks_schafer`: Considering the other already existing implementations, I think the inversion should not occur inside the function and rather be done manually by the user beforehand.
- `_rsp_findpeaks_schafer`: More a question to @DominiqueMakowski: All other peak finders use the `_rsp_findpeaks_outliers` and `_rsp_findpeaks_sanitize` functions to their output. I think with this implementation, it would make sense to skip on these though. Or is it important that the rsp always starts with a trough? I think it might make sense to have a `clean` parameter and let the output be more flexible. This is for a different thread though.

# Signal
- `signal_quality`: I see that you renamed the `signal_quality` file to `signal_templatequality` and created a separate file for the `signal_ibiquality`. Usually this repository looks at keeping everything in one and separating them out based on the `method` keyword.
- `signal_ibiquality`: The concept of ibi seems quite specific to place in the signal category. I understand why you did it (since it works for both ppg and ecg), but is there a more general term for it that would fit better into the signal category? Like looking at a cycles (inter cycle interval?)
- `signal_ibiquality`: The "See Also" is missing your new `ppg_quality`.
- `signal_ibiquality`: Should we make tolerance a parameter here? In the Ho et al. (2025) paper you cite it mentions "In this study we used a tolerance window of 50ms, which differs from the 150ms used in previous studies..." so shorter windows may also be good to have.
- `signal_cyclesegment`: This is really nice!

# The Errors
- To me the code bot seems to be correct. Taking a look at `_calc_template_morph`, it seems to already do the indexing / filtering that you are attempting to do there, so just passing `beat_inds` already does the job. Not sure why it was `beat_inds[0:-1]` before.

# General
- I tried out all the new functions, and they seem to work well in my limited testing. Nice!
- All new implemented features and paths of code should be covered with unit tests (let me know if I should write these for you or if I can help here)
- Code quality looks very nice and great documentation!

Thanks you again for your hard work!

Johannes

peterhcharlton commented:
I have now addressed the areas where the tests were failing, and hopefully the code in commit [89fccb3](https://github.com/neuropsychology/NeuroKit/pull/1131/commits/89fccb3f4be82c3401115a1a3480893c78fec0bc) addresses these issues.

DerAndereJohannes commented:
Thanks! I think it looks good to me. I have added some additional tests (which I added to your already created tests). They mainly exist to check for regressions if something changes and checking the different methods based on known values of what it should be (e.g., empty signals or a perfect sine wave should have perfect quality etc.).

Regarding the "potential error", I think you have solved it and it should be fine. Also thanks for finding some other places where errors could have occured like in the `signal_period` command. That was definitely an oversight.

If there is one thing I would like to ask you to do before pushing this to `dev` are if the results of the `ici` method looks like they make sense relative to their input signal. I will point to these in an additional comment.

P.S. Excuse me that I highjacked your request to fix the line length formatting for the entire repository (I just ran the regular black command on it).

DerAndereJohannes commented on tests/tests_signal.py:
```diff
@@ -458,3 +458,157 @@ def test_signal_surrogate():
+                if method == 'templatematch':
+                    assert np.isclose(np.mean(quality), 1, atol=1e-2)
+                elif method == 'dissimilarity':
+                    assert np.isclose(np.mean(quality), 0, atol=1e-3)
```
Here you can see the different tests to test the results from `signal_quality` on the different physiological sensor data. I have used the numpy isclose command to see if they are good quality. The largest question I have for you is if the `ici` results look reasonable (also plot the signal and check) since they have quite a lenient tolerance of up to 0.1 difference. Thanks!

DerAndereJohannes replied:
thank you so much for looking into it with more detail and finding additional considerations that I did not see before.

I think there is no perfect solution to the question of whether the start/end should stay 0 or 1. With smaller signals, the start and end being 0 will make it more difficult to get higher scores. The reverse can be said if we have a signal with very few cycles in the middle and nothing else on the edges, leading to very high scores. 

I feel like the quality metric would also have to take into account how much of the signal actually holds cycles and state it based off of that? But i guess for such periodic signals such as ECG / PPG this should not be a problem. In this case I added "The resulting quality signal is 1-padded, with all samples before the first cycle and after the last cycle set to 1." to the documentation to make the decision clear.

I think we are good to merge!
